src/message_bus.zig

```diff
@@ -73,60 +191,413 @@ pub const MessageBus = struct {
+        bytes_received: usize = 0,
+        incoming_header: Header = undefined,
+        incoming_message: *Message = undefined,
+
+        /// This completion is used for all send operations.
+        send_completion: IO.Completion = undefined,
+        /// The queue of messages to send to the client.
+        /// Empty unless peer == .client
+        send_queue: FIFO(Envelope) = .{},
+        bytes_sent: usize = 0,
+
+        /// Given a newly accepted fd, start receiving messages on it.
+        /// Callbacks will be continously re-registered until close() is
+        /// called and the connection is terminated.
+        fn receive_messages(self: *Connection, fd: os.socket_t) void {
+            assert(self.fd == -1);
+            assert(self.peer == .none);
+            self.fd = fd;
+            self.peer = unknown;
+            self.recv_header();
+        }
+
+        fn recv_header(self: *Connection) void {
+            // The fd may have been closed due to an error while sending.
+            if (self.fd == -1) return;
+            self.message_bus.io.recv(
+                *ReplicaConnection,
+                self,
+                complete_recv_header,
+                &self.completion,
+                mem.asBytes(&self.incoming_header)[self.bytes_received..],
+                os.MSG_NOSIGNAL,
+            );
+        }
+
+        fn complete_recv_header(self: *Connection, completion: *IO.Completion, result: RecvError!usize) void {
+            self.bytes_received += result catch |err| {
+                // TODO: maybe don't need to close on *every* error
+                log.err("error receiving data from {}: {}", .{ self.peer, err });
+                self.close();
+                return;
+            };
+
+            if (self.bytes_received < @sizeOf(Header)) {
+                // The header has not yet been fully received.
+                self.recv_header();
+                return;
+            }
+            assert(self.bytes_received == @sizeOf(Header));
+
+            if (!self.incoming_header.valid_checksum()) {
+                log.err("invalid checksum on header received from {}", .{self.peer});
+                self.close();
+                return;
+            }
+
+            assert(self.peer != .none);
+            if (self.peer == .unknown) {
+                if (self.incoming_header.command == .request) {
+                    self.peer = .{ .client = self.incoming_header.client };
+                } else {
+                    self.message_bus.replica_connections += 1;
+                    self.peer = .{ .replica = self.incoming_header.replica };
+                }
+            }
+
+            self.incoming_message = self.message_bus.create_message(self.incoming_header.size) catch unreachable;
+            self.incoming_message.references += 1;
+            self.incoming_message.header.* = self.incoming_header;
+            self.recv_body();
+        }
+
+        fn recv_body(self: *Connection) void {
```

Comment:
Elsewhere we use `data` instead of `body` e.g. `Header.valid_checksum_data()`. But if you agree `body` is better, we can do a quick grep/replace and switch to that everywhere instead of `data`. We probably should.

Reply:
I do think `body` is more clear when referring to the non-header part of a message. `data` makes me wonder if the header is included or not.

Reply:
100% agree. I'll rename to `body` in `vr.zig`

Reply:
Done: https://github.com/coilhq/tigerbeetle/commit/67c91bc8845853475352d9397629d33dbff421d1

This updates two lines in `beta/src/message_bus.zig`.